float32 cx
float32 cy
float32 ctheta